Simplified/NYPLNetworkQueue.swift

```diff
@@ -0,0 +1,173 @@
+  private static let sqlHeader = Expression<Data?>("requestHeader")
+  private static let sqlRetries = Expression<Int>("retryCount")
+  
+  
+  // MARK: - Public Functions
+  
+  class func addRequest(_ libraryID: Int,
+                        _ updateID: String?,
+                        _ requestUrl: URL,
+                        _ method: HTTPMethodType,
+                        _ parameters: Data?,
+                        _ headers: [String : String]?) -> Void
+  {
+    // Serialize Data
+    let urlString = requestUrl.absoluteString
+    let methodString = method.rawValue
+    
+    let headerData: Data?
+    if headers != nil {
+      headerData = NSKeyedArchiver.archivedData(withRootObject: headers!)
+    } else {
+      headerData = nil
+    }
+    
+    guard let db = startDatabaseConnection() else { return }
+    
+    // Get or create table
+    do {
+      try db.run(sqlTable.create(ifNotExists: true) { t in
+        t.column(sqlID, primaryKey: true)
+        t.column(sqlLibraryID)
+        t.column(sqlUpdateID)
+        t.column(sqlUrl)
+        t.column(sqlMethod)
+        t.column(sqlParameters)
+        t.column(sqlHeader)
+        t.column(sqlRetries)
+      })
+    } catch {
+      Log.error(#file, "SQLite Error: Could not create table")
+      return
+    }
+    
+    // Update (not insert) if uniqueID and libraryID match existing row in table
+    let query = sqlTable.filter(sqlLibraryID == libraryID && sqlUpdateID == updateID)
+                        .filter(sqlUpdateID != nil)
+    
+    do {
+      // Update row
+      if try db.run(query.update(sqlParameters <- parameters, sqlHeader <- headerData)) > 0 {
+        Log.info(#file, "SQLite Row Updated - Success")
+        
+      // Insert new row
+      } else {
+        do {
+          try db.run(sqlTable.insert(sqlLibraryID <- libraryID, sqlUpdateID <- updateID, sqlUrl <- urlString, sqlMethod <- methodString, sqlParameters <- parameters, sqlHeader <- headerData, sqlRetries <- 0))
+          Log.info(#file, "SQLite Row Added - Success")
+        } catch {
+          Log.error(#file, "SQLite Error: Could not update table")
+        }
+      }
+    } catch {
+      Log.error(#file, "SQLite Error: Could not update queue")
+    }
+  }
+  
+  class func retryQueue()
+  {
+    guard let db = startDatabaseConnection() else { return }
+    
+    do {
+      for row in try db.prepare(sqlTable) {
+        self.retry(db: db, requestRow: row)
+      }
+    } catch {
+      Log.error(#file, "SQLite Error accessing table")
+    }
+  }
+  
+  
+  // MARK: - Private Functions
+  
+  private class func retry(db: Connection, requestRow: Row)
+  {
+    if (Int(requestRow[sqlRetries]) > QueueMaxRetryCount) {
+      deleteRow(db: db, id: Int(requestRow[sqlID]))
+      Log.info(#file, "Removing after \(Int(requestRow[sqlRetries])) retries")
+      return
+    }
+    
+    do {
+      let ID = Int(requestRow[sqlID])
+      let newValue = Int(requestRow[sqlRetries]) + 1
```

Comment:
> To your first point, the queue should only be retrying when NYPLReachability sends a notification that the host is reachable.

Ah, okay! Perhaps were good as-is then. 

> I do like adding some kind of time limit though. Something like 30 days?

I think that might be okay! Perhaps something like "discard after **at least** M days **and at least** N failures" would cover both cases? I wouldn't want something to be dropped because it was tried three times during an hour of a temporary server outage and I also wouldn't want something to be dropped just because the user's device has been off for a month and we never got to try again.